### PR TITLE
Refactor general sdk code and expand tests

### PR DESCRIPTION
**Pull request by max-vogler**
### Summary

Improved `NameValue`, `RequestBuilder`, `RequestUtils` and `VisualRecognition`

### CODE REVIEW

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 1]
```diff
@@ -238,5 +301,51 @@ public void testCustomUserAgent() throws InterruptedException {
...
+  public void testNoAuthentication() {
+    new SpeechToText().getModels();
+  }
+
+  /**
+   * Tests request with skipped authentication
+   * @throws InterruptedException the interrupted exception
+   */
+  @Test
+  public void testSkipAuthentication() throws InterruptedException {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722166938/issues/source?bid=3303699&fileBranchId=3347904#l318)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 2]
```diff
@@ -62,6 +86,45 @@ private RecordedRequest checkRequest() throws InterruptedException {
   }
 
   /**
+   * Checks the credential constructors of all services.
+   */
+  @Test
+  public void testConstructors() {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722166938/issues/source?bid=3303699&fileBranchId=3347904#l92)

File: src/main/java/com/ibm/watson/developer_cloud/visual_recognition/v3/VisualRecognition.java [hunk 3]
```diff
@@ -92,16 +92,15 @@ public VisualRecognition(String versionDate) {
...
+   * Instantiates a new Visual Recognition V3 service with a given API key.
    *
    * @param versionDate The version date (yyyy-MM-dd) of the REST API to use. Specifying this value
    *        will keep your API calls from failing when the service introduces breaking changes.
-   * @param username the username
-   * @param password the password
+   * @param apiKey the api key
    */
-  public VisualRecognition(String versionDate, String username, String password) {
+  public VisualRecognition(String versionDate, String apiKey) {
```
Comment (max-vogler) [inline src/main/java/com/ibm/watson/developer_cloud/visual_recognition/v3/VisualRecognition.java]: New unit tests for the constructors found that this constructor is **always** throwing an Exception. In the spirit of *fail-fast programming* I introduced breaking API changes to remove the obsolete constructor.

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 4]
```diff
@@ -238,5 +301,51 @@ public void testCustomUserAgent() throws InterruptedException {
...
+  public void testNoAuthentication() {
+    new SpeechToText().getModels().execute();
+  }
+
+  /**
+   * Tests request with skipped authentication
+   * @throws InterruptedException the interrupted exception
+   */
+  @Test
+  public void testSkipAuthentication() throws InterruptedException {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722420316/issues/source?bid=3303699&fileBranchId=3347904#l318)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java (hunk 2 above)
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722420316/issues/source?bid=3303699&fileBranchId=3347904#l92)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 5]
```diff
@@ -238,5 +302,57 @@ public void testCustomUserAgent() throws InterruptedException {
     assertTrue(request.getHeader(HttpHeaders.USER_AGENT).endsWith("foo-bar"));
     service.setDefaultHeaders(null);
   }
-  
+
+  /**
+   * Tests request without authentication.
+   */
+  @Test
+  public void testNoAuthentication() {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722996208/issues/source?bid=3303699&fileBranchId=3347904#l310)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 6]
```diff
@@ -238,5 +302,57 @@ public void testCustomUserAgent() throws InterruptedException {
...
+      // success!
+    }
+  }
+
+  /**
+   * Tests request with skipped authentication
+   * @throws InterruptedException the interrupted exception
+   */
+  @Test
+  public void testSkipAuthentication() throws InterruptedException {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722996208/issues/source?bid=3303699&fileBranchId=3347904#l325)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 7]
```diff
@@ -238,5 +302,57 @@ public void testCustomUserAgent() throws InterruptedException {
...
+  /**
+   * Tests request without authentication.
+   */
+  @Test
+  public void testNoAuthentication() {
+    try {
+      new SpeechToText().getModels();
+      throw new AssumptionViolatedException("createServiceCall() did not throw an IllegalArgumentException, " +
+        "even though no authentication has been specified.");
+    } catch (IllegalArgumentException e) {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [Avoid empty catch blocks](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722996208/issues/source?bid=3303699&fileBranchId=3347904#l315)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 8]
```diff
@@ -238,5 +302,57 @@ public void testCustomUserAgent() throws InterruptedException {
...
-  
+
+  /**
+   * Tests request without authentication.
+   */
+  @Test
+  public void testNoAuthentication() {
+    try {
+      new SpeechToText().getModels();
+      throw new AssumptionViolatedException("createServiceCall() did not throw an IllegalArgumentException, " +
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [In JUnit4, use the @Test(expected) annotation to denote tests that should throw exceptions](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722996208/issues/source?bid=3303699&fileBranchId=3347904#l313)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java [hunk 9]
```diff
@@ -62,6 +87,45 @@ private RecordedRequest checkRequest() throws InterruptedException {
   }
 
   /**
+   * Checks the credential constructors of all services.
+   */
+  @Test
+  public void testConstructors() {
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2722996208/issues/source?bid=3303699&fileBranchId=3347904#l93)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java (hunk 5 above)
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2724346036/issues/source?bid=3303699&fileBranchId=3347904#l310)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java (hunk 6 above)
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2724346036/issues/source?bid=3303699&fileBranchId=3347904#l325)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java (hunk 7 above)
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [Avoid empty catch blocks](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2724346036/issues/source?bid=3303699&fileBranchId=3347904#l315)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java (hunk 8 above)
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [In JUnit4, use the @Test(expected) annotation to denote tests that should throw exceptions](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2724346036/issues/source?bid=3303699&fileBranchId=3347904#l313)

File: src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java (hunk 9 above)
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/developer_cloud/service/GenericServiceTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/watson-developer-cloud/java-sdk/file/2724346036/issues/source?bid=3303699&fileBranchId=3347904#l93)

**Pull request merged**